File: 2023/lang-rust/src/day01.rs
use std::fs;

const MAPPING: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

fn part_1(input: &str) -> u32 {
    input
        .lines()
        .map(|line: &str| {
            let mut digits = line.chars().filter_map(|c| c.to_digit(10));
            let f = digits.next().unwrap();
            let l = digits.last().unwrap_or(f);

            f * 10 + l
        })
        .sum()
}

fn part_2(input: &str) -> u32 {
    input
        .lines()
        .map(|line: &str| {
            let mut x = line.chars();
            let mut digit_1 = None;
            let mut digit_2 = None;

            while x.as_str() != "" {
                let string = x.as_str();
                let char = x.next().unwrap();

                if char.is_ascii_digit() {
                    digit_2 = char.to_digit(10);
                    digit_1 = digit_1.or(digit_2);
                } else {
                    for (idx, str) in MAPPING.iter().enumerate() {
                        if string.starts_with(str) {
                            digit_2 = Some(idx as u32 + 1);
                            digit_1 = digit_1.or(digit_2);
                        }
                    }
                }
            }

            digit_1.unwrap() * 10 + digit_2.unwrap()
        })
        .sum()
}

pub fn solve() {
    let input = fs::read_to_string("../input/day01.txt").expect("Missing");

    println!("Day 01, Part 1 = {}", part_1(&input));
    println!("Day 01, Part 2 = {}", part_2(&input));
}

#[cfg(test)]
mod tests {
    use super::super::*;
    use std::fs;

    #[test]
    fn test_part_1_example() {
        let input = fs::read_to_string("../input/day01-p1t.txt").expect("Missing");
        assert_eq!(day01::part_1(&input), 142);
    }

    #[test]
    fn test_part_1() {
        let input = fs::read_to_string("../input/day01.txt").expect("Missing");
        assert_eq!(day01::part_1(&input), 55108);
    }

    #[test]
    fn test_part_2() {
        let input = fs::read_to_string("../input/day01.txt").expect("Missing");
        assert_eq!(day01::part_2(&input), 56324);
    }
}

File: 2023/lang-php/src/Day01.php
<?php

class Day01
{
    const MAPPING = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine'];

    public static function part1(string $input): int
    {
        $lines = preg_replace(subject: trim($input), pattern: '/([A-Za-z])/', replacement: '');
        $lines = explode(string: $lines, separator: \PHP_EOL);
        $output = 0;

        foreach ($lines as $line) {
            $f = substr($line, 0, 1);
            $l = substr($line, -1);

            $output += intval($f . $l);
        }

        return $output;
    }

    public static function part2(string $input): int
    {
        $lines = explode(string: trim($input), separator: \PHP_EOL);
        $output = 0;

        foreach ($lines as $line) {
            $length = strlen($line);
            $offset = 0;
            $f = null;
            $l = null;

            for ($offset = 0; $offset < $length; $offset++) {
                if (is_numeric($line[$offset])) {
                    $l = intval($line[$offset]);
                    $f = $f ?? $l;
                } else {
                    $string = substr($line, $offset);

                    foreach (self::MAPPING as $idx => $value) {
                        if (str_starts_with($string, $value)) {
                            $l = $idx + 1;
                            $f = $f ?? $l;
                        }
                    }
                }
            }

            $output += intval($f . $l);
        }

        return $output;
    }
}

$input = file_get_contents(filename: '../input/day01.txt');
echo sprintf("Day 01, Part 01 = %d", Day01::part1($input)) . \PHP_EOL;
echo sprintf("Day 01, Part 02 = %d", Day01::part2($input)) . \PHP_EOL;
